### PR TITLE
nixos: expose user session environment during activation

### LINKED ISSUE #1399: "onchange" command doesn't work when using as nixos module.
### Issue description

Hi! when I use as nixos module I have a problem.
Looks like most modules work as excepted but when I'm using xsession.windowManager.xmonad.config and edit the file, doesn't auto recompile and restart as excepted.
I looked xmonad.nix code and looks like it uses `home.file.".xmonad/xmonad.hs".onChange` to detect file change and recompile/restart.
I also have <code>".xmonad/xmobar.hs".onChange =</code> in my home.nix that I personally use but doesn't work.
so it's not a module specific issue.
Sorry if this is my misunderstanding and there are option or something to make work.
I don't have idea what is wrong.

#### Maintainer CC
I will not mention because I don't think it's a module specific isuue. and I don't have idea what is causing the issue.
Edit: I think actually it is a module. maybe files.nix? but there are not meta.maintainers.

#### Technical details
```
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/d43v6bx7r6fcaq3fbbfd5mwh1f5s7rmg-bash-interactive-4.4-p23-devcopying path '/nix/store/d43v6bx7r6fcaq3fbbfd5mwh1f5s7rmg-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"` - host os: `Linux 5.7.9, NixOS, 20.03.2611.bb8f0cc2279 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.2611.bb8f0cc2279"`
 - channels(btw): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### LINKED ISSUE #2112: nixos: graphical session variables are not available during activation
Related to #1399.

When using the NixOS module, graphical session variables like `DISPLAY` are not set when the activation script runs, which breaks some `onChange` hooks ([xmonad](https://github.com/nix-community/home-manager/blob/63af2d3e4cac0a84a87b72be0135f5bcec5a9d5d/modules/services/window-managers/xmonad.nix#L146-L149), [xresources](https://github.com/nix-community/home-manager/blob/63af2d3e4cac0a84a87b72be0135f5bcec5a9d5d/modules/xresources.nix#L88-L90)).

This could be solved by using a user service (`systemd.user.services` or `systemd.userActivationScripts`) to activate the Home Manager generation, and relying on `DISPLAY` and consorts being imported into the user's systemd manager's environment.

NixOS does not allow to configure those services per-user, so we'd need to have a single script that either
  - does something like `case $USER in ${user1}) ${user1Activation};; ${user2}) ${user2Activation};; ...; esac` but this could get ugly and potentially inefficient if there are many users
  - or simply runs `home-manager-generation` if that command is available, after we make sure to add each user's activation package to `users.users.<name>.packages`. I like this solution as it also provides a command to the user to reset their HM config.

Please tell me if you see potential issues with this.

cc @rycee 

EDIT: a first issue I see is that now we can't set `RequireMountsFor` or `restartTriggers` per-user.

EDIT: an easier solution might be to eval `systemctl --user show-environment` in the activation script, or running it with `systemd-run --user`, but those seem to require hardcoding `XDG_RUNTIME_DIR`.

### PR DESCRIPTION
**Pull request by ncfavier**
If the user has a running systemd session, source their environment from the systemd manager and export a few variables in order to allow activation scripts to reload applications on the fly.

The list of variables to export is arbitrary and could be extended in the future.

Fixes #1399, fixes #2112.

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

### CODE REVIEW

**Comment by ncfavier** [comment]
Made sure to only export the listed variables, and added `XDG_RUNTIME_DIR` and `DBUS_SESSION_BUS_ADDRESS` to the list.

**Comment by zseri** [comment]
I would put the added part into a separate script, and pass that script the generation directory as an argument. This way both problems would be solved, and we would have the benefit that the wrapper script could be shared between multiple HM generations.

e.g.:
```nix
{
#...
          # The activation script is run by a login shell to make sure
          # that the user is given a sane Nix environment.
          ExecStart = let
            systemctl =
              "XDG_RUNTIME_DIR=\${XDG_RUNTIME_DIR:-/run/user/$UID} systemctl";
            exportedVariables = [
              "XDG_RUNTIME_DIR"
              "DBUS_SESSION_BUS_ADDRESS"
              "DISPLAY"
              "XAUTHORITY"
            ];
            tweakEnv = pkgs.writeScript "tweak-env" ''
              #! ${pkgs.runtimeShell} -el
              # If the user is logged in, import variables from their current
              # session environment.
              eval "$(${systemctl} --user show-environment 2> /dev/null |
                      sed -En '/^(${
                        concatStringsSep "|" exportedVariables
                      })=/s/^/export /p')"
              exec "$1/activate"
            '';
          in "${tweakEnv} ${usercfg.home.activationPackage}/activate";
#...
}
```

File: nixos/default.nix [hunk 1]
```diff
@@ -140,8 +140,25 @@ in {
...
           # that the user is given a sane Nix environment.
-          ExecStart = pkgs.writeScript "activate-${username}" ''
+          ExecStart = let
+            systemctl =
+              "XDG_RUNTIME_DIR=\${XDG_RUNTIME_DIR:-/run/user/$UID} systemctl";
+            exportedVariables = [
+              "XDG_RUNTIME_DIR"
+              "DBUS_SESSION_BUS_ADDRESS"
+              "DISPLAY"
+              "XAUTHORITY"
```
Comment (berbiche) [inline nixos/default.nix:150]:
```suggestion
            exportedVariables = [
              "XDG_RUNTIME_DIR"
              "DBUS_SESSION_BUS_ADDRESS"
              "DISPLAY"
              "XAUTHORITY"
              "WAYLAND_DISPLAY"
```

File: nixos/default.nix [hunk 2]
```diff
@@ -140,8 +140,25 @@ in {
...
+            # If the user is logged in, import variables from their current
+            # session environment.
+            eval "$(${systemctl} --user show-environment 2> /dev/null |
+                    sed -En '/^(${
+                      concatStringsSep "|" exportedVariables
+                    })=/s/^/export /p')"
+
             echo Activating home-manager configuration for ${username}
             exec ${usercfg.home.activationPackage}/activate
           '';
```
Comment (zseri) [inline nixos/default.nix:164]:
```suggestion
            tweakEnv = pkgs.writeScript "tweak-env" ''
              #! ${pkgs.runtimeShell} -el
              # If the user is logged in, import variables from their current
              # session environment.
              eval "$(${systemctl} --user show-environment 2> /dev/null |
                      sed -En '/^(${
                        concatStringsSep "|" exportedVariables
                      })=/s/^/export /p')"
              exec "$1/activate"
            '';
          in "${tweakEnv} ${usercfg.home.activationPackage}";
```

File: nixos/default.nix [hunk 3]
```diff
@@ -140,8 +140,25 @@ in {
...
+              "DISPLAY"
+              "XAUTHORITY"
+            ];
+          in pkgs.writeScript "activate-${username}" ''
             #! ${pkgs.runtimeShell} -el
+
+            # If the user is logged in, import variables from their current
+            # session environment.
+            eval "$(${systemctl} --user show-environment 2> /dev/null |
+                    sed -En '/^(${
```
Comment (zseri) [inline nixos/default.nix:158]: may I ask where this `sed` comes from? `systemPackages`?

Comment (ncfavier) [reply nixos/default.nix:158]: [Yes](https://github.com/NixOS/nixpkgs/blob/dbe5054806f2a57e023eac29ff9284cd3f15c128/nixos/modules/config/system-path.nix#L26).

**Review by zseri — APPROVED** [review]

**Comment by zseri** [comment]
This should be rebased on top of the master branch and the conflict fixed (by ignoring the change on master).

**Comment by ncfavier** [comment]
@rycee I would appreciate a review on this

**Comment by rycee** [comment]
Thanks! Merged to master now. I made a few minor changes and have tried it out for a while without issues.

**Comment by ncfavier** [comment]
Thanks!

**Pull request merged**